src/hotspot/cpu/x86/stubGenerator_x86_64_aes.cpp

```diff
@@ -3177,4 +3290,493 @@ void StubGenerator::aesgcm_encrypt(Register in, Register len, Register ct, Regis
   __ movq(rax, pos);
 }
 
+void StubGenerator::gfmul_avx2(XMMRegister GH, XMMRegister HK) {
+  const XMMRegister T1 = xmm1;
+  const XMMRegister T2 = xmm2;
+  const XMMRegister T3 = xmm3;
+
+  __ vpclmulqdq(T1, GH, HK, 0x11); // %%T1 = a1*b1
+  __ vpclmulqdq(T2, GH, HK, 0x00); // %%T2 = a0*b0
+  __ vpclmulqdq(T3, GH, HK, 0x01); // %%T3 = a1*b0
+  __ vpclmulqdq(GH, GH, HK, 0x10); // %%GH = a0*b1
+  __ vpxor(GH, GH, T3, Assembler::AVX_128bit);
+
+  __ vpsrldq(T3, GH, 8, Assembler::AVX_128bit); // shift-R %%GH 2 DWs
+  __ vpslldq(GH, GH, 8, Assembler::AVX_128bit); // shift-L %%GH 2 DWs
+
+  __ vpxor(T1, T1, T3, Assembler::AVX_128bit);
+  __ vpxor(GH, GH, T2, Assembler::AVX_128bit);
+
+  //first phase of the reduction
+  __ movdqu(T3, ExternalAddress(ghash_polynomial_reduction_addr()), r15 /*rscratch*/);
+  __ vpclmulqdq(T2, T3, GH, 0x01);
+  __ vpslldq(T2, T2, 8, Assembler::AVX_128bit); // shift-L %%T2 2 DWs
+
+  __ vpxor(GH, GH, T2, Assembler::AVX_128bit); // first phase of the reduction complete
+  //second phase of the reduction
+  __ vpclmulqdq(T2, T3, GH, 0x00);
+  __ vpsrldq(T2, T2, 4, Assembler::AVX_128bit); // shift-R %%T2 1 DW (Shift-R only 1-DW to obtain 2-DWs shift-R)
+
+  __ vpclmulqdq(GH, T3, GH, 0x10);
+  __ vpslldq(GH, GH, 4, Assembler::AVX_128bit); // shift-L %%GH 1 DW (Shift-L 1-DW to obtain result with no shifts)
+
+  __ vpxor(GH, GH, T2, Assembler::AVX_128bit); // second phase of the reduction complete
+  __ vpxor(GH, GH, T1, Assembler::AVX_128bit); // the result is in %%GH
+}
+
+void StubGenerator::generateHtbl_8_block_avx2(Register htbl, Register rscratch) {
+  const XMMRegister HK = xmm6;
+
+  __ movdqu(HK, Address(htbl, 0));
+  __ movdqu(xmm1, ExternalAddress(ghash_long_swap_mask_addr()), rscratch);
+  __ vpshufb(HK, HK, xmm1, Assembler::AVX_128bit);
+
+  __ movdqu(xmm11, ExternalAddress(ghash_polynomial_addr()), rscratch);
+  __ movdqu(xmm12, ExternalAddress(ghash_polynomial_two_one_addr()), rscratch);
+  // Compute H ^ 2 from the input subkeyH
+  __ movdqu(xmm2, xmm6);
+  __ vpsllq(xmm6, xmm6, 1, Assembler::AVX_128bit);
+  __ vpsrlq(xmm2, xmm2, 63, Assembler::AVX_128bit);
+  __ movdqu(xmm1, xmm2);
+  __ vpslldq(xmm2, xmm2, 8, Assembler::AVX_128bit);
+  __ vpsrldq(xmm1, xmm1, 8, Assembler::AVX_128bit);
```

Comment:
You could save a few bytes here:
```suggestion
  __ vpsrlq(xmm1, xmm6, 63, Assembler::AVX_128bit);
  __ vpsllq(xmm6, xmm6, 1, Assembler::AVX_128bit);
  __ vpslldq(xmm2, xmm1, 8, Assembler::AVX_128bit);
  __ vpsrldq(xmm1, xmm1, 8, Assembler::AVX_128bit);
```